### PR TITLE
Moved some APIs around to make Result<T, E> covariant

### PR DESCRIPTION
**Pull request by lnewson**
Related Ticket: N/A

Description of Changes:
* Removed .is from Result.ts
* Added .is to Results.ts (note the s)
* Added a test to make sure that results are now covariant
* Enabled `strictFunctionTypes` for katamari

Note: This is a follow up to the work done for Optionals in https://github.com/tinymce/tinymce/pull/6884 by @ashfordneil.

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Review by tiny-james — APPROVED** [review]

**Review by JamesToohey — APPROVED** [review]

**Review by ltrouton — APPROVED** [review]

File: modules/katamari/src/test/ts/atomic/api/data/ResultCovarianceAssertionTest.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+interface Cat extends Animal {
+  readonly length: number;
+}
+
+// https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)
+//
+// We need to make sure that Result<T, E> is, and remains, covariant with
+// respect to T and E.
+describe('atomic.katamari.api.data.ResultCovarianceAssertionTest', () => {
```
Comment (ashfordneil) [inline modules/katamari/src/test/ts/atomic/api/data/ResultCovarianceAssertionTest.ts:19]: Might as well merge the pair of these into a single `CovarianceAssertionTest` and get Optional and Result in the same group — 99% of the code is the same.

Also, does this test still pass with strict mode on (I know everything else in katamari breaks, just want to make sure this file in particular doesn't have any errors so we don't get bitten like last time)

File: modules/katamari/src/test/ts/module/ephox/katamari/test/arb/ArbDataTypes.ts [hunk 2]
```diff
@@ -7,10 +7,10 @@ import { Result } from 'ephox/katamari/api/Result';
 
 type Arbitrary<A> = fc.Arbitrary<A>;
 
-export const arbResultError = <A, E> (arbE: Arbitrary<E>): Arbitrary<Result<A, E>> =>
+export const arbResultError = <A = never, E = any> (arbE: Arbitrary<E>): Arbitrary<Result<A, E>> =>
```
Comment (ashfordneil) [inline modules/katamari/src/test/ts/module/ephox/katamari/test/arb/ArbDataTypes.ts:10]: No need for an `any` here
```suggestion
export const arbResultError = <E, A = never> (arbE: Arbitrary<E>): Arbitrary<Result<A, E>> =>
```

File: modules/katamari/src/main/ts/ephox/katamari/api/Result.ts [hunk 3]
```diff
@@ -103,7 +97,7 @@ const value = <T, E = any>(o: T): Result<T, E> => {
   };
 };
 
-const error = <T = any, E = any>(message: E): Result<T, E> => {
+const error = <T = never, E = any>(message: E): Result<T, E> => {
```
Comment (ashfordneil) [inline modules/katamari/src/main/ts/ephox/katamari/api/Result.ts:100]: Awesome to have this `any` removed — we should try to get the other one on this line in the next slack time PR

**Review by ashfordneil — APPROVED** [review]

**Review by HAFRMO — APPROVED** [review]

File: modules/katamari/src/test/ts/module/ephox/katamari/test/arb/ArbDataTypes.ts (hunk 2 above)
Comment (lnewson) [reply modules/katamari/src/test/ts/module/ephox/katamari/test/arb/ArbDataTypes.ts:10]: I'm not doing this, as it requires switching the order which is going to be confusing as the order for results is always `<Value, Error>`.

File: modules/katamari/src/test/ts/atomic/api/data/ResultCovarianceAssertionTest.ts (hunk 1 above)
Comment (lnewson) [reply modules/katamari/src/test/ts/atomic/api/data/ResultCovarianceAssertionTest.ts]: Have merged the 2 tests into a single file. Also yes, I double checked and there's no errors in this test when strict mode is enabled.

**Pull request merged**